karabo_data/geometry2.py

```diff
@@ -500,6 +493,128 @@ def to_distortion_array(self):
+        pixel. If not, the coordinates are for the first corner of the pixel
+        (the one nearest the [0, 0] corner of the tile in data space).
+        """
+        out = np.zeros(self.expected_data_shape + (3,), dtype=np.float64)
+
+        # Prepare some arrays to use inside the loop
+        pixel_ss_coord, pixel_fs_coord = np.meshgrid(
+            np.arange(0, self.frag_ss_pixels, dtype=np.float64),
+            np.arange(0, self.frag_fs_pixels, dtype=np.float64),
+            indexing='ij'
+        )
+
+        # This is where the DSSC subclass shifts odd rows by half a pixel
+        self._adjust_pixel_coords(pixel_ss_coord, pixel_fs_coord)
+
+        if centre:
+            # A pixel is from n to n+1 in each axis, so centres are at n+0.5.
+            pixel_ss_coord += 0.5
+            pixel_fs_coord += 0.5
+
+        for m, mod in enumerate(self.modules, start=0):
+            for t, tile in enumerate(mod, start=0):
+                corner_x, corner_y, corner_z = tile.corner_pos * self.pixel_size
+                ss_unit_x, ss_unit_y, ss_unit_z = tile.ss_vec * self.pixel_size
+                fs_unit_x, fs_unit_y, fs_unit_z = tile.fs_vec * self.pixel_size
+
+                # Calculate coordinates of each pixel's first corner
+                # 2D arrays, shape: (64, 128)
+                pixels_x = (
+                        corner_x
+                        + pixel_ss_coord * ss_unit_x
+                        + pixel_fs_coord * fs_unit_x
+                )
+                pixels_y = (
+                        corner_y
+                        + pixel_ss_coord * ss_unit_y
+                        + pixel_fs_coord * fs_unit_y
+                )
+                pixels_z = (
+                        corner_z
+                        + pixel_ss_coord * ss_unit_z +
+                        + pixel_fs_coord * fs_unit_z
+                )
+
+                # Which part of the array is this tile?
+                tile_ss_slice, tile_fs_slice = self._tile_slice(t)
+
+                # Insert the data into the array
+                out[m, tile_ss_slice, tile_fs_slice, 0] = pixels_x
+                out[m, tile_ss_slice, tile_fs_slice, 1] = pixels_y
+                out[m, tile_ss_slice, tile_fs_slice, 2] = pixels_z
+
+        return out
+
+    def data_coords_to_physical(self, module_no, slow_scan, fast_scan):
+        """Convert coordinates in data space to physical space
+
+        module_no, fast_scan and slow_scan should all be numpy arrays of the
+        same shape. module_no should hold integers, starting from 0,
+        so 0: Q1M1, 1: Q1M2, etc.
+
+        slow_scan and fast_scan describe positions within that module.
+        They may hold floats for sub-pixel positions. In both, 0.5 is the centre
+        of the first pixel.
+
+        Returns an array of similar shape with an extra dimension of length 3,
+        for (x, y, z) coordinates in metres.
+        """
```

Comment:
I do not understand this method and what should be the inputs :(

Reply:
I'll try to clarify it tomorrow. But briefly, it's for things like the output of the robust peak finder: it gives peak locations in pixel coordinates within the input data - e.g. 31.2 pixels in the slow-scan direction, 10.78 pixels fast scan. This method can convert those into coordinates in physical space (x, y, z).